website/docs/r/nsxt_distributed_firewall_rule.html.markdown

```diff
@@ -0,0 +1,117 @@
+---
+layout: "vcd"
+page_title: "VMware Cloud Director: vcd_nsxt_distributed_firewall_rule"
+sidebar_current: "docs-vcd-resource-nsxt-distributed-firewall-rule"
+description: |-
+  The Distributed Firewall allows user to segment organization virtual data center entities, such as
+  virtual machines, based on virtual machine names and attributes. 
+---
+
+# vcd\_nsxt\_distributed\_firewall\_rule
+
+The Distributed Firewall allows user to segment organization virtual data center entities, such as
+virtual machines, based on virtual machine names and attributes. 
+
+~> There is a different resource
+[`vcd_nsxt_distributed_firewall`](/providers/vmware/vcd/latest/docs/resources/nsxt_distributed_firewall)
+that can manage all firewall rules in one resource. One should use **only one of**
+`vcd_nsxt_distributed_firewall` or `vcd_nsxt_distributed_firewall_rule` as they will **override each
+other**.
+
+## Example Usage
+
+```hcl
+data "vcd_vdc_group" "existing" {
+  org  = "my-org" # Optional, can be inherited from Provider configuration
+  name = "main-vdc-group"
+}
+
+data "vcd_nsxt_network_context_profile" "cp1" {
+  context_id = vcd_vdc_group.test1.id
+  name       = "CTRXICA"
+  scope      = "SYSTEM"
+}
+
+resource "vcd_nsxt_distributed_firewall_rule" "r1" {
+  org          = "my-org"
+  vdc_group_id = vcd_vdc_group.test1.id
+
+  name        = "rule1"
+  action      = "ALLOW"
+  description = "description"
+
+  source_ids           = [vcd_nsxt_ip_set.set1.id, vcd_nsxt_ip_set.set2.id]
+  destination_ids      = [vcd_nsxt_security_group.g1-empty.id, vcd_nsxt_security_group.g2.id]
+  app_port_profile_ids = [vcd_nsxt_app_port_profile.p1.id, data.vcd_nsxt_app_port_profile.WINS.id, data.vcd_nsxt_app_port_profile.FTP.id]
+}
+
+resource "vcd_nsxt_distributed_firewall_rule" "r2" {
+  org          = "{{.Org}}"
+  vdc_group_id = vcd_vdc_group.test1.id
+
+  # Specifying a particular ID of other firewall rule will ensure that the current one is placed above
+  above_rule_id = vcd_nsxt_distributed_firewall_rule.r1.id
+
+  name        = "rule4"
+  action      = "ALLOW"
+  ip_protocol = "IPV6"
+  direction   = "OUT"
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `org` - (Optional) The name of organization to use, optional if defined at provider level. Useful
+  when connected as sysadmin working across different organisations.
+* `above_rule_id` - (Optional) ID of another `vcd_nsxt_distributed_firewall_rule` entry will ensure
+  that newly created rule is positioned above the references rule. **Note.** By default, new rule
```

Comment:
Oh, there seems to be some confusion here with the naming. I thought that `above_rule_id` means a pointer to the rule **above** the current one, but it seems it's not. In which case, I feel we should make the current name more deterministic (I know, it's a hard naming challenge)...

CC: @dataclouder @adambarreiro 

Also, small typo:

```suggestion
  that newly created rule is positioned above the referenced rule. **Note.** By default, new rule
```

Reply:
Typo fixed.

As for the name itself - I chose it from NSX-V world, because this was what the NSX-V API had https://registry.terraform.io/providers/vmware/vcd/latest/docs/resources/nsxv_firewall_rule#above_rule_id

Reply:
That's actually good, as it means we have lived with it and our users have survived the naming ambiguity. The issue is I can't think of an alternative (better) name yet :) ...

So, I'd like to suggest a simplifaction to the field description at least (changes in bold):

> ID of **an existing** `vcd_nsxt_distributed_firewall_rule` entry **above which**
>   **the** newly created **firewall** rule **will be positioned**.

Reply:
p.s. I think we don't have to be fully consistent with NSX-V, so here are some alternative name suggestions for consideration. I would like at least to open the idea bucket. All with the goal of removing any ambiguity, which is especially useful when you're reading HCLs written by someone else without knowing the details. So:

1. `place_above_rule_id`
2. `position_above_rule_id`

CC: @dataclouder @adambarreiro

Reply:
I have adjusted the proposed comment. Open to field name changes.

Reply:
A final call for comments on https://github.com/vmware/terraform-provider-vcd/pull/1076#discussion_r1252975276 from @dataclouder @adambarreiro @adezxc ^^ Which name do you prefer: leave `above_rule_id` or change to `place_above_rule_id` or `position_above_rule_id`?

Reply:
IMHO, what I like from `above_rule_id` is that describes better the state, like "This rule is above the rule with ID", vs `place...` or `position...` which look more like an imperative action, "Place this rule above the rule with ID". As Terraform tends to describe a "desired state" vs actions, I'd prefer `above_rule_id` even though it may be confusing at first (I think the confusion gets cleared once the docs are read).

But let's see what others think :)

Reply:
It could also be that the confusion is generated by the fact that "above" in this context can mean "This rule is above the specified" or "The above rule is this one".
Maybe choosing a more exact word, like `in_preference_to_rule_id` or something similar.

Reply:
Yes, @adambarreiro that was exactly the confusion I hit at the start of review. To contrast, both "above rule ID" and "bellow rule ID" may be interpreted exactly the same in this context.

Reply:
I think I prefer sticking with `above_rule_id`, all argument names will probably cause some kind of confusion, so docs will be needed anyway. I agree with @adambarreiro's point that `place_above_rule_id` is too explicit and not really declarative.